server/routes.go

```diff
@@ -1010,6 +1010,44 @@ func streamResponse(c *gin.Context, ch chan any) {
 	})
 }
 
+// contextLimitPrompt builds a prompt to send to a running model while trimming it to fit the template into the max context length
+func contextLimitPrompt(ctx context.Context, prompts []string, ctxLen int) (string, error) {
+	if len(prompts) == 0 {
+		return "", nil
+	}
+
+	prompt := ""
+	promptLen := 0
+
+	// always process the last prompt
+	lastPrompt := prompts[len(prompts)-1]
+	enc, err := loaded.runner.Encode(ctx, lastPrompt)
+	if err != nil {
+		return "", err
+	}
+	prompt = lastPrompt
+	promptLen = len(enc)
+
+	if len(prompts) == 1 {
+		return prompt, nil
+	}
+
+	// iterate over the rest of the prompts in reverse order until we reach the max context length
+	for i := len(prompts) - 2; i >= 0; i-- {
```

Comment:
I'm not sure the separation is warranted. Why not reverse the slice and iterate as normal for all messages? The special case for the first (last) message could be an additional check:

```go
var prompt string
for i, message := range slices.Reverse(prompts) {
  enc , err := loaded.runner.Encode(ctx, message)
  // check err

  if i != 0 && promptLen + len(enc) > loaded.NumCtx {
    break
  }

  prompt += message + prompt
  promptLen += len(enc)
}
```